File: ex_4.cpp
#include <iostream>
#include <cstring>
#include <unistd.h>

#define BUFFER_SIZE 25
#define Read_end 0
#define write_end  1
using namespace std; // Adding namespace std

int main(void){
    char write_msg[BUFFER_SIZE] = "greet";
    char read_msg[BUFFER_SIZE];
    int fd[2];
    pid_t pid;

    // Creating a pipe
    if (pipe(fd) == -1) {
        cerr << "Pipe failed" << endl;
        return 1;
    }

    // Creating a child process
    pid = fork();
    if (pid < 0) {
        cerr << "Fork failed" << endl;
        return 1;
    }
    // Parent process
    if (pid > 0) {
        close(fd[Read_end]); // Close the read end of the pipe in the parent
        // Writing message to the pipe
        write(fd[write_end], write_msg, strlen(write_msg) + 1);
        close(fd[write_end]); // Close the write end of the pipe in the parent
    }
    // Child process
    else {
        close(fd[write_end]); // Close the write end of the pipe in the child
        // Reading message from the pipe
        read(fd[Read_end], read_msg, BUFFER_SIZE);
        cout << "Read: " << read_msg << endl;
        close(fd[Read_end]); // Close the read end of the pipe in the child
    }
    return 0;
}

File: ex_4.py
import os
import multiprocessing

def child_process(message, pipe_out):
    pipe_out.send(message.encode())
    pipe_out.close()

def parent_process(pipe_in):
    message = pipe_in.recv().decode()
    print("Message received from child process:", message)

if __name__ == "__main__":
    # Get the message from the user
    message = input("Enter the message: ")

    # Creating a pipe
    pipe_in, pipe_out = multiprocessing.Pipe()

    # Creating a child process
    child = multiprocessing.Process(target=child_process, args=(message, pipe_out))
    child.start()
    child.join()

    # Close the write end of the pipe in the parent
    pipe_out.close()

    # Read the message from the pipe in the parent
    parent_process(pipe_in)

